### ISSUE TITLE
Code audits

### ISSUE THREAD

**Issue by nh2** [Reviewable/Reviewable#452]
I do code audits for companies from time to time, and it would be great if Reviewable supported those so that I can recommend it to them for that purpose.

A code audit is different from a normal PR/commit review in that it doesn't review a _change_, but a current _state_ of the repository at the time you choose. So there's no diff, no green/red highlights, and only one side in the side-by-side diff. Comments can be made anywhere, and typically the whole repo or a whole sub-directory is under review, and one often has to jump between files to convince oneself that some code is correct (so a directory navigator would very useful).

(A common hack to implement code audits with standard review tools is to make a commit that deletes all code, and another commit that adds it back, and then review only the latter. That kind of works, but it's ugly, makes everything green in the review (a bit of a silly point, but in practice it makes code harder to read), and you typically don't want to merge this change into your history, which means you cannot merge the bit that you officially reviewed.)

After the initial phase of the code audit, the author or maintainer of the code may make requested changes. At this point the audit turns into a real review, and Reviewable's awesome revision sliders would come into action again. But again, the starting point of the review chain would be a _state_ of the repo, not a _change_.

I imagine the interface for starting an audit in Revieable that you pick a starting revision and a target branch. The starting revision would be the initial state to review, and it has its own revision in the revision slider, and when you select only that, you get the full tree to review. All commits between `starting revision..target branch` would be added as revisions to the right of the starting revision in the revision slider (or as a single revision if you don't enable per-commit reviews). As the target branch moves forward (e.g. in reaction to audit commits), new entries get added to the revision slider as usual.

Of the Review tools I know, only Atlassian Crucible has support for this, and it does it pretty well, and companies like it because it does. But its reviewing usability is not as good as Reviewable, so I'd prefer Reviewable over it a lot if it could do it.

Audits are common in companies: many get consultants for external general purpose code audits, security audits, legal audits. I also wish they were common for open source projects, and it would be great if that could be done in public, e.g. a public vetting of a crypto library by a well-known cryptographer. Or simply if you want to use some library for your project, but you'd prefer to read all its code before you start relying on it, and have Reviewable help you with tracking your progress on that.

It would be great if Reviewable could support this first-class.

**Comment by pkaminski** [comment]
Thanks for taking the time to do a nice write-up Niklas!  A similar feature was brought up before in #51, and #222 is also peripherally related.  Most of my doubts from the earlier issue remain:  this could prove to be a distraction, complicate the UI, and I'm not convinced the code audit market is large enough to warrant catering to it (pointers to data appreciated).

From the implementation point of view, Reviewable is tied tightly to GitHub's pull requests so adding the ability to create a review directly against a branch would require a significant amount of work.  The most likely path would be to allow arbitrary files to be added to an existing review, but even that has some unpleasant ramifications into Reviewable's infrastructure (which assumes it has control of the file list).

@mikesokolsky, what do you think?

**Comment by mikesokolsky** [comment]
I'm definitely familiar with the frustration of wanting to do a code-review style audit of an entire library and only being able to get partial diffs, so I'm personally a +1 on this feature.  Thanks for the detailed write-up as well @nh2.

I does seem like it will be a pretty involved change to do as a full feature.  The suggested work-around means it is functionally possible right now, if very messy.  As we go through the new review API I'll keep an eye out for possible ways to get to a 90% solution with minimal effort.

**Comment by nh2** [comment]
I don't have any market demand data, but here's a comment from a colleague today who didn't know about Reviewable yet:

> It would be awesome if there was some software that would like you snapshot github repo, make comments from all aspects, add in automated linting, etc all in one interface

**Comment by ShawnTalbert** [comment]
Darn, I was afraid I might get redirected here. We aren't doing audits, we just do code review after logical commit rather than a blocker to commit  (aka PR).

OK well I'll try to watch for progress on this issue but I will need to look for another product that supports our lightweight workflow for now.

**Comment by pkaminski** [comment]
Yeah, I realize it's not a perfect fit for your use case, but it shares the same "diff a branch/commit against an arbitrary target" mechanic.  Reviewable is definitely aimed at a precommit review workflow at the moment.  If you're doing postcommit you'll need another tool; if you're doing precommit without PRs, how come?

**Comment by ShawnTalbert** [comment]
we're doing postcommit, by design.
